### PR TITLE
Replace "volatile" keyword with a macro to read shared data

### PR DESCRIPTION
**Pull request by michaelcahill**
This is intended to stimulate some discussion.  I don't find "volatile" easy to maintain because we have to know when declaring a field how it will be used.  That said, calling out "shared reads" isn't necessarily easy either, and would be easy to forget.

refs WT-1938

### CODE REVIEW

**Comment by michaelcahill** [comment]
@keithbostic, @sueloverso, @agorrod what do you think of something like this?

I think what we're doing now (a mix of volatile with some barriers) is the worst of both worlds: since volatile is at the declaration site, we can't know from looking at a loop on its own whether or not it is safe.  I'd prefer either explicit barriers in every loop that is polling shared state, or something like this change to what we're doing now.

Thoughts?

File: src/include/cache.h [hunk 1]
```diff
@@ -94,18 +94,12 @@ struct __wt_cache {
...
 	WT_DATA_HANDLE
 		*evict_file_next;	/* LRU next file to search */
 
 	/*
-	 * Sync/flush request information.
-	 */
-	volatile uint64_t sync_request;	/* File sync requests */
-	volatile uint64_t sync_complete;/* File sync requests completed */
-
-	/*
```
Comment (michaelcahill) [inline src/include/cache.h]: I went through everything marked volatile: these fields aren't being used.

**Comment by keithbostic** [comment]
I'm on the fence (sorry about that), on this one. Thoughts in no particular order:

1. Do we mix barriers and volatile on the same fields, or do we use volatile in some places, and barriers in others? I didn't see any place we did both, are there some?

1. There's no `WT_SHARED_WRITE` call: were there no places we depend on volatile preventing the compiler from skipping "unnecessary" writes?

1. The change adds restrictions to the compiler, it can no longer do out-of-order execution across the read, which might lessen code optimization. Technically, it also removes restrictions from the compiler, it can skip unnecessary writes, if we care.

1. I don't think `WT_SHARED` captures the idea, would you consider `WT_FORCED`?

1. If we don't have `volatile` next to the declaration, is there more risk someone will see the value in a structure and simply use it without realizing the need for special care?

1. Exchanging volatile for a barrier might allow us to more easily port to systems with weaker cache guarantees.

Not really thought through, but maybe we could consider more structure?

We could obscure the names of the volatile structure fields, and automatically generate a set of inline functions to read/write them. Code using the inline functions documents how the field is being accessed at any particular place, and there's no chance of it being used without at least some thought.

**Comment by michaelcahill** [comment]
@keithbostic, I don't disagree with your comments, but I'm reluctant to do more magic with field declarations.

What I'm really concerned about is lurking bugs that only show up in optimized builds under load, or after some innocuous code change.  I'd be happy to have a "POLLING_LOOP" macro that adds a barrier to the beginning of any loop that is relying on some shared state changing.

I note that in the Linux kernel, they rely on their yield operations working as barriers.  I'd be happy enough with that, too, but we have some loops with `goto` in various places that neither `__wt_yield` or `WT_PAUSE`.  Every tight loop should at least do the latter...

**Comment by sueloverso** [comment]
I like Michael's suggestion of a `WT_POLLING_WHILE` macro.  It is something that is easy to remember, as I'm one likely to forget!  We have a similar precedent with the `_FOREACH_BEGIN/END` macros.

**Comment by michaelcahill** [comment]
We're not going to take this change.  Revisit a "polling loop" macro in the future.

**Pull request closed without merging**